20250524-R-sentry-jwt-oidc.md

```diff
@@ -0,0 +1,138 @@
+
+**Trade-offs:**
+- Increases Sentry configuration and deployment complexity
+- Exposes unauthenticated endpoints (user must secure as needed)
+
+**Advantages:**
+- Enables federated identity scenarios and cloud provider integration
+- No breaking changes to existing APIs
+
+**Disadvantages:**
+- Additional configuration required for Sentry
+- Endpoint security is the user's responsibility
+
+## Implementation Details
+
+### Design
+
+- Sentry configuration extended for JWT signing key, algorithm, JWKS (auto-generated if not provided)
+- JWT issued alongside x.509 SVID in `SignCertificate` response
+- Audience claim set from app ID configuration (default: trust domain)
+- JWKS endpoint for public key distribution
+- Optional OIDC discovery server serves `.well-known/openid-configuration` and `jwks.json`
+- OIDC server exposed via Kubernetes Service (user responsible for exposure and protection)
+- JWT rotation uses same mechanism as x.509 rotation
+- SPIFFE package extended to support JWT-SVIDs
+- daprd runtime and components can use JWT for cloud federation via token provider
+
+#### Proposed SignCertificateResponse
+
+```go
+// SignCertificateResponse is returned by Sentry's SignCertificate API.
+type SignCertificateResponse struct {
+    // Existing x.509 SVID fields
+	WorkloadCertificate []byte
+	TrustChainCertificates [][]byte
+	ValidUntil             *timestamppb.Timestamp
+
+	// JWT token for authentication. Always provided by Sentry if configured to issue JWTs.
+	Jwt *string
+}
+```
+
+#### Proposed token provider for components-contrib
+
+```go
+// Token provider function to retrieve JWT from SPIFFE context
+tokenProvider := func(ctx context.Context) (string, error) {
+    // New method on spiffecontext to fetch JWT SVID
+	tknSource, ok := spiffecontext.JWTFrom(ctx)
+	if !ok {
+		return "", fmt.Errorf("failed to get JWT SVID source from context")
+	}
+	jwt, err := tknSource.FetchJWTSVID(ctx, jwtsvid.Params{
+		Audience: AzureADTokenExchangeAudience,
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get JWT SVID: %w", err)
+	}
+	return jwt.Marshal(), nil
+}
+```
+
+### Proposed Sentry Configuration Flags
+
+**JWT-related flags:**
+- `--jwt-enabled` (bool): Enable JWT token issuance by Sentry
+- `--jwt-key-filename` (string): JWT signing key filename
+- `--jwks-filename` (string): JWKS (JSON Web Key Set) filename
+- `--jwt-issuer` (string): Issuer value for JWT tokens (no issuer if empty)
+- `--jwt-signing-algorithm` (string): Algorithm for JWT signing, must be supported by signing key
+- `--jwt-key-id` (string): Key ID (kid) for JWT signing
+
+**OIDC-related flags:**
+- `--oidc-http-port` (int): Port for the OIDC HTTP server (disabled if 0)
+- `--oidc-jwks-uri` (string): Custom URI for external JWKS access
+- `--oidc-path-prefix` (string): Path prefix for all OIDC HTTP endpoints
+- `--oidc-domains` (string slice): Allowed domains for OIDC HTTP endpoint requests
```

Comment:
What does this do? Check the SNI headers for incoming well known http requests? 🤔

Reply:
TLS is terminated here, but because you're exposing an unauthenticated HTTP server this just allows you to restrict the hostname in the request to ensure the requests are coming via the expected route e.g. `oidc.myissuer.net` and not via some internal or alternative domain.